stl/inc/ranges

```diff
@@ -6837,6 +6837,602 @@ namespace ranges {
+                          || (!(bidirectional_range<_RangeTypes> && ...) && (common_range<_RangeTypes> && ...)) //
+                          || ((random_access_range<_RangeTypes> && ...) && (sized_range<_RangeTypes> && ...));
+
+        template <typename _CallbackType>
+        constexpr auto _Create_tuple_transform_lambda(_CallbackType&& _Callback) noexcept {
+            const auto _Created_lambda = [&_Callback]<typename... _ViewTupleTypes>(_ViewTupleTypes&&... _View_tuples) {
+                return tuple<invoke_result_t<_CallbackType&, _ViewTupleTypes>...>{
+                    _STD invoke(_Callback, _STD forward<_ViewTupleTypes>(_View_tuples))...};
+            };
+
+            return _Created_lambda;
+        }
+
+        template <typename _CallbackType, typename _TupleType>
+        constexpr auto _Tuple_transform(_CallbackType&& _Callback, _TupleType&& _Tuple) noexcept(
+            noexcept(_STD apply(_Zip::_Create_tuple_transform_lambda(_STD forward<_CallbackType>(_Callback)),
+                _STD forward<_TupleType>(_Tuple)))) /* strengthened */ {
+            return _STD apply(_Zip::_Create_tuple_transform_lambda(_STD forward<_CallbackType>(_Callback)),
+                _STD forward<_TupleType>(_Tuple));
+        }
+
+        template <typename _CallbackType>
+        constexpr auto _Create_tuple_for_each_lambda(_CallbackType&& _Callback) noexcept {
+            const auto _Created_lambda = [&_Callback]<typename... _ViewTupleTypes>(_ViewTupleTypes&&... _View_tuples) {
+                (_STD invoke(_Callback, _STD forward<_ViewTupleTypes>(_View_tuples)), ...);
+            };
+
+            return _Created_lambda;
+        }
+
+        template <typename _CallbackType, typename _TupleType>
+        constexpr void _Tuple_for_each(_CallbackType&& _Callback, _TupleType&& _Tuple) noexcept(
+            noexcept(_STD apply(_Zip::_Create_tuple_for_each_lambda(_STD forward<_CallbackType>(_Callback)),
+                _STD forward<_TupleType>(_Tuple)))) /* strengthened */ {
+            _STD apply(_Zip::_Create_tuple_for_each_lambda(_STD forward<_CallbackType>(_Callback)),
+                _STD forward<_TupleType>(_Tuple));
+        }
+
+        template <size_t _DesiredOffset, size_t _FirstIndex, size_t... _OtherIndices>
+        struct _Offset_index_sequence_solver {
+            static constexpr auto _Solved_index_sequence{
+                _Offset_index_sequence_solver<_DesiredOffset, _OtherIndices...>::_Solved_index_sequence};
+        };
+
+        template <size_t _DesiredOffset, size_t... _OtherIndices>
+        struct _Offset_index_sequence_solver<_DesiredOffset, _DesiredOffset, _OtherIndices...> {
+            static constexpr auto _Solved_index_sequence{index_sequence<_DesiredOffset, _OtherIndices...>{}};
+        };
+
+        template <size_t _OriginalSequenceSize, size_t _Offset>
+            requires(_Offset <= _OriginalSequenceSize)

```

Comment:
Could this be a `static_assert`, or perhaps an `_STL_INTERNAL_STATIC_ASSERT` (which is active only for our test code, not for users, when we're worried about logic errors in the library instead of user misuse)? We generally use concepts to switch between implementations, or constrain constructors etc. to exist only when they would be valid, but here this is just an internal function with a precondition, and if the `requires` isn't satisfied, there are no other overloads that would be called.